Add error message option for `belongs_to_required_by_default` on the associations

aditya-kapoor opened the pull request:
We have a multi lingual app which is hosted on the different subdomains. This app is backed an English only Admin panel where we manage different models. 

The admin panel when accessed through the different locales (subdomains) results in missing translations for the models which have `belongs_to` associations in them. We have handled the cases for the attributes having well defined validation rules using generic english messages but I think this is another area where we felt there is a need for the ability to customize the error messages for the belongs_to associations.

aditya-kapoor commented:
@eileencodes any updates?

jagdeepsingh commented:
@aditya-kapoor can you include an example of how it behaves with/without this implementation?

jagdeepsingh commented on activerecord/lib/active_record/associations/builder/belongs_to.rb:
```diff
@@ -147,7 +151,7 @@ def self.define_validations(model, reflection)
 
       if required
-        model.validates_presence_of reflection.name, message: :required
+        model.validates_presence_of reflection.name, message: (validation_message.presence || :required)
```
I know, at most one validation is defined for `belongs_to` (and `has_one`) relation, and that is of presence only, so, this code will work. But the option you specified to be able to define a custom validation message i.e. `:message` does not correctly represent what validation is this message for, e.g. in case there are more than one validations possible (i can't think of any now, but future is uncertain). Just a point from me.

jagdeepsingh replied:
May be `belongs_to :user, message: { presence: 'Custom error message' }` or something better.

aditya-kapoor replied:
I believe the `:presence` key is not required here because the only validation implied now is `required`. So here it is intended to do.

```
   belongs_to :user, message: 'must be present.'
```

aditya-kapoor commented:
@rafaelfranca @eileencodes any update on this?

eileencodes reviewed (changes requested):
Can you document this change?

aditya-kapoor committed:
Add error message option for `belongs_to_required_by_default` on the associations.

aditya-kapoor committed:
Merge branch 'master' into add-required-msg-option

aditya-kapoor commented:
@eileencodes done.

1. Added an entry in CHANGELOG.md.
2. Added the description in the API documentation along with example.

Can I also add this in the Guides too?

sikachu commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -1492,6 +1492,9 @@ def has_many(name, scope = nil, **options, &extension)
         #   +:inverse_of+ to avoid an extra query during validation.
+        # [:message]
+        #   Specifies the error message for +required+ option.
+        #   NOTE: If <tt>required</tt> is set to <tt>true</tt> then only this option will work.
```
Do you mind revise this wording a bit? I don't think _only this option will work_ is the right description for this flag.

```ruby
#  NOTE: This option will be used only when `:required` flag is set to `true`.
```

Alternatively, we can also remove this note as line 1496 already made it clear that this works in conjunction with `:required`.

sikachu commented on activerecord/lib/active_record/associations/builder/belongs_to.rb:
```diff
@@ -145,6 +145,10 @@ def self.define_validations(model, reflection)
       end
 
+      if reflection.options.key?(:message)
+        validation_message = reflection.options.delete(:message)
```
What do you think about naming this key `:validation_message` or something similar? I believe this will improve the readability and really reflect what the usage of this value will be.

```ruby
belongs_to :user, required: true, validation_message: "should be set"
```

sikachu commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -1649,6 +1656,7 @@ def has_one(name, scope = nil, **options)
         #   belongs_to :comment, touch: true
         #   belongs_to :company, touch: :employees_last_updated_at
         #   belongs_to :user, optional: true
+        #   belongs_to :user, message: 'must be present.'
```
What do you think about showing this example with both `:required` and `:message` (`:validation_message`) set, since those two goes together?

adityakapoor-homebell replied:
👍

sikachu commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -1492,6 +1492,9 @@ def has_many(name, scope = nil, **options, &extension)
         #   This will validate the association itself, not the id. You can use
         #   +:inverse_of+ to avoid an extra query during validation.
+        # [:message]
+        #   Specifies the error message for +required+ option.
```
`required` => `:required`

sikachu reviewed (changes requested):
I have a few feedbacks regarding the naming. I hope you don't mind.

kaspth commented:
I don't see how this helps your multi-lingual app.

How does `belongs_to :person, message: 'ohøj'` work when the `I18n.locale` is `:en` instead of the example `:da` that I used? E.g. load time messages can't change on runtime locale changes.

I also think that having a `message:` argument on `belongs_to` seems to far outside its responsibility.

Consider turning `belongs_to_required_by_default` off and using `validates_presence_of` instead.

How are Active Record's I18n support failing short here? https://guides.rubyonrails.org/i18n.html#translations-for-active-record-models
